sense/stick/stick.go

```diff
@@ -1,73 +1,115 @@
 package stick
 
 import (
-	"errors"
-	"fmt"
-	"io/ioutil"
 	"os"
-	"path/filepath"
 	"strings"
+	"syscall"
+	"unsafe"
 )
 
-// Joystick events
-var (
-	Right = 0
-	Left  = 1
-	Up    = 2
-	Down  = 3
-	Enter = 4
+const (
+	Enter = 28
+	Up    = 103
+	Left  = 105
+	Right = 106
+	Down  = 108
 )
 
-// the joy stick
-var joyStickDevice string
+type Device struct {
+	f      *os.File
+	Events chan Event
+}
 
-func init() {
-	var err error
-	joyStickDevice, err = getDevice("Raspberry Pi Sense HAT Joystick")
+func Open(name string) (*Device, error) {
+	f, err := os.Open(name)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
+
+	d := &Device{f, make(chan Event, 4)}
+
+	go d.pollEvents()
+
+	return d, nil
 }
 
-// ReadEvent from the joystick.
-func ReadEvent() int {
-	file, err := os.Open(joyStickDevice)
-	if err != nil {
-		panic(err)
-	}
+func (d *Device) Close() error {
+	var err error
 
-	data := make([]byte, 100)
-	count, err := file.Read(data)
-	if err != nil {
-		panic(err)
+	if d.f != nil {
+		err = d.f.Close()
+		d.f = nil
 	}
-	fmt.Printf("read %d bytes, %q\n", count, data[:count])
 
-	return count
+	return err
 }
 
-func getDevice(name string) (string, error) {
-	matches, err := filepath.Glob("/sys/class/input/event*")
-	if err != nil {
-		return "", err
-	}
+func (d *Device) Name() string {
+	var str [256]byte
+
+	ioctl(d.f.Fd(), _EVIOCGNAME(256), uintptr(unsafe.Pointer(&str[0])))
+
+	return strings.TrimRight(string(str[:]), "\x00")
+}
 
-	for _, dir := range matches {
-		b, err := ioutil.ReadFile(filepath.Join(dir, "device/name"))
+func (d *Device) pollEvents() {
+	defer close(d.Events)
+
+	var e Event
+
+	size := int(unsafe.Sizeof(e))
+	buf := make([]byte, size*2)
+
+	for {
+		n, err := d.f.Read(buf)
 		if err != nil {
-			continue
+			return
 		}
-		fbName := strings.TrimSpace(string(b))
-		if fbName == name {
-			dev := filepath.Join("/dev/input", filepath.Base(dir))
 
-			return dev, nil
+		events := (*(*[1<<27 - 1]Event)(unsafe.Pointer(&buf[0])))[:n/size]
+
+		for i := range events {
+			if e := events[i]; e.Type == 0x01 && e.Value != 0 {
+				d.Events <- e
+			}
 		}
 	}
-	return "", errStickDeviceNotFound
 }
 
-// errors
-var (
-	errStickDeviceNotFound = errors.New("stick device not found")
+// Event represents a generic input event.
+type Event struct {
+	Time  syscall.Timeval
+	Type  uint16
+	Code  uint16
+	Value int32
+}
+
+func ioctl(fd, name, v uintptr) error {
+	_, _, errno := syscall.RawSyscall(syscall.SYS_IOCTL, fd, name, v)
+	if errno == 0 {
+		return nil
+	}
+
+	return errno
+}
+
+func _EVIOCGNAME(len int) uintptr {
+	return _IOC(_IOC_READ, 'E', 0x06, len)
+}
+
+const (
+	_IOC_WRITE     = 0x1
```

Comment:
I believe there is some benefit of matching the constants used in the Linux kernel (`drivers/input/evdev.c` and `include/uapi/asm-generic/ioctl.h`), but that might be non-obvious for people that just want to use this package.